src/storage/src/sink/kafka.rs

```diff
@@ -601,655 +630,253 @@ impl KafkaSinkState {
                                 value: Some(&ts_bytes),
                             }));
 
-                            let size_for_stats =
-                                u64::cast_from(record.payload.as_ref().map_or(0, |p| p.len()))
-                                    + u64::cast_from(record.key.as_ref().map_or(0, |k| k.len()));
                             total_size_for_stats += size_for_stats;
 
-                            s.send(record).await;
+                            s.send(record).await?;
                             sink_statistics.inc_messages_staged_by(1);
                             sink_statistics.inc_bytes_staged_by(size_for_stats);
-
-                            // advance to the next repetition of this row, or the next row if all
-                            // repetitions are exhausted
-                            repeat_counter += 1;
-                            if repeat_counter == encoded_row.count {
-                                repeat_counter = 0;
-                                s.metrics.rows_queued.dec();
-                            }
+                            s.metrics.rows_queued.dec();
                         }
 
                         // Flush to make sure that errored messages have been properly retried before
                         // sending progress records and commit transactions.
-                        s.flush().await;
+                        s.flush().await?;
 
                         // We don't count this record as part of the message count in user-facing
                         // statistics.
-                        s.send_progress_record(*ts).await;
+                        s.send_progress_record(upper.clone()).await?;
+
+                        info!("{}: committing transaction for {}", s.name, upper.pretty());
+                        s.producer
+                            .retry_on_txn_error(|p| p.commit_transaction())
+                            .await?;
 
-                        info!("Committing transaction for {:?}", ts,);
-                        s.halt_on_err(
-                            s.producer
-                                .retry_on_txn_error(|p| p.commit_transaction())
-                                .await,
-                        )
-                        .await;
                         sink_statistics.inc_messages_committed_by(count_for_stats);
                         sink_statistics.inc_bytes_committed_by(total_size_for_stats);
 
-                        s.flush().await;
+                        s.flush().await?;
 
-                        // sanity check for the continuous updating
-                        // of the write frontier below
-                        s.assert_progress(ts);
-                        progress_update.replace(ts.clone());
+                        pending_batches.retain(|batch| !batch.is_empty());
+                        lower.clone_from(&upper);
+                        *write_frontier.borrow_mut() = upper;
```

Comment:
Sure! This property is pervasive in materialize and is enforced by the storage controller for all collections. If the since of a collection ever became equal to the upper then that collection would become unreadable.

For each export the controller knows its current `write_frontier`, which is what gets set here with the *pristine* upper frontier we just sinked to kafka. This update travels back to the controller and eventually we reach the `update_write_frontiers` method: https://github.com/MaterializeInc/materialize/blob/89ebf6f41ca9b8a06329994f0f22868e19b9f7f1/src/storage-controller/src/lib.rs#L1385-L1393

Here ^ the `read_capability` of the sink is updated based on its read policy and the write frontier we just reported. The read policy that we use for exports is the `LagWritesBy` policy which does the `saturating_sub(1)` thing you mentioned. Here is where we set the read policy: https://github.com/MaterializeInc/materialize/blob/89ebf6f41ca9b8a06329994f0f22868e19b9f7f1/src/storage-controller/src/lib.rs#L1054

And this is how that policy converts an upper frontier into a since frontier:

https://github.com/MaterializeInc/materialize/blob/89ebf6f41ca9b8a06329994f0f22868e19b9f7f1/src/storage-client/src/controller.rs#L548-L566

So circling back to what `update_write_frontier` does, it updates this new read capability based on the upper and calls `update_read_capabilities`. That method will propagate the update since requirement to the dependency of the sink, advancing our read hold.

Reply:
Cool! Fair to summarize that: while we're removing one minus-one from the sink frontier, that was redundant with a second minus-one that already happens in the controller, so we can remove one entirely without causing any problems?

Reply:
That is correct!